src/cljam/io/twobit/reader.clj

```diff
@@ -1,125 +1,133 @@
-                 :big (.readInt rdr))]
-    {:name name-str
-     :offset offset}))
-
-(defn- read-sequence-header!
-  [^DataInput rdr endian]
-  (let [read-fn (case endian :little lsb/read-int :big #(.readInt ^DataInput %))
-        dna-size (read-fn rdr)
-        n-block-count (read-fn rdr)
-        n-block-starts (doall (repeatedly n-block-count #(inc (read-fn rdr))))
-        n-block-sizes (doall (repeatedly n-block-count #(read-fn rdr)))
-        mask-block-count (read-fn rdr)
-        mask-block-starts (doall (repeatedly mask-block-count #(inc (read-fn rdr))))
-        mask-block-sizes (doall (repeatedly mask-block-count #(read-fn rdr)))
-        zero (read-fn rdr)]
-    {:len dna-size
-     :ambs (mapv vector n-block-starts n-block-sizes)
-     :masks (mapv vector mask-block-starts mask-block-sizes)
-     :header-offset (+ 16 (* n-block-count 4 2) (* mask-block-count 4 2))}))
-
-(def ^:const twobit-to-str
+  (close [this]))
+
+(defrecord ChromHeader [ambs masks ^long header-offset])
+
+(defrecord Chrom [name ^int len ^int offset ^int index header])
+
+(defn- ^TreeMap read-header-block! [^ByteBuffer buf]
+  (let [n-blocks (.getInt buf)
+        starts (doto (.slice buf) (.order (.order buf)))
+        _ (.position buf (+ (.position buf) (* Integer/BYTES n-blocks)))
+        m (TreeMap.)]
+    (dotimes [_ n-blocks]
+      (.put m (unchecked-inc-int (.getInt starts)) (.getInt buf)))
+    (assert (= (.size m) n-blocks))
+    m))
+
+(defn- read-sequence-header! [buf]
+  (let [amb-blocks (read-header-block! buf)
+        mask-blocks (read-header-block! buf)]
+    (ChromHeader.
+     amb-blocks
+     mask-blocks
+     (* Integer/BYTES
+        (+ 4 (* (.size amb-blocks) 2) (* (.size mask-blocks) 2))))))
+
+(defn- read-file-index! [^ByteBuffer buf ^long n-seqs]
+  (let [m (HashMap.)
+        ba (byte-array 255)]
+    (dotimes [i n-seqs]
+      (let [chr-len (Byte/toUnsignedInt (.get buf))
+            _ (.get buf ba 0 chr-len)
+            chr (String. ba 0 chr-len)
+            offset (.getInt buf)
+            _ (.mark buf)
+            _ (.position buf offset)
+            len (.getInt buf)
+            _ (.reset buf)
+            header (delay
+                    (let [buf' (.duplicate buf)]
+                      (.order buf' (.order buf))
+                      (.position buf' (+ offset Integer/BYTES))
+                      (read-sequence-header! buf')))]
+        (.put m chr (Chrom. chr len offset i header))))
+    m))
+
+(def ^:private ^"[[C" twobit-to-str
   (let [table "TCAG"]
-    (mapv
-     (fn [j] (let [i (byte (- j 128))
-                   n4 (bit-and i 2r11)
-                   n3 (bit-and (unsigned-bit-shift-right i 2) 2r11)
-                   n2 (bit-and (unsigned-bit-shift-right i 4) 2r11)
-                   n1 (bit-and (unsigned-bit-shift-right i 6) 2r11)]
-               (str (.charAt table n1) (.charAt table n2) (.charAt table n3) (.charAt table n4))))
-     (range 256))))
+    (->> 256
+         range
+         (mapv
```

Comment:
It's a little bit too trivial, but there's no need to use `mapv` rather than `map` any more?